lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -2250,84 +2256,32 @@ constructTransaction
-        handleMissingAssetName mb = case mb ^. #assetName of
-            Nothing -> mb {assetName = Just $ ApiT nullTokenName}
-            Just _ -> mb
-    let mintingBurning' = fmap handleMissingAssetName <$> mintingBurning
-    let retrieveAllCosigners = foldScript (:) []
-    let wrongMintingTemplate (ApiMintBurnData (ApiT scriptTempl) _ _) =
-            isLeft (validateScriptOfTemplate RecommendedValidation scriptTempl)
-            || length (retrieveAllCosigners scriptTempl) /= 1
-            || (L.any (/= Cosigner 0)) (retrieveAllCosigners scriptTempl)
-    when
-        ( isJust mintingBurning' &&
-          L.any wrongMintingTemplate (NE.toList $ fromJust mintingBurning')
-        ) $ liftHandler $ throwE ErrConstructTxWrongMintingBurningTemplate
-
-    let assetNameTooLong = \case
-            (ApiMintBurnData _ (Just (ApiT (UnsafeTokenName bs))) _) ->
-                BS.length bs > tokenNameMaxLength
-            _ ->
-                error "tokenName should be nonempty at this step"
-    when
-        ( isJust mintingBurning' &&
-          L.any assetNameTooLong (NE.toList $ fromJust mintingBurning')
-        ) $ liftHandler $ throwE ErrConstructTxAssetNameTooLong
-
-    let assetQuantityOutOfBounds
-            (ApiMintBurnData _ _ (ApiMint (ApiMintData _ amt))) =
-            amt <= 0 || amt > unTokenQuantity txMintBurnMaxTokenQuantity
-        assetQuantityOutOfBounds
-            (ApiMintBurnData _ _ (ApiBurn (ApiBurnData amt))) =
-            amt <= 0 || amt > unTokenQuantity txMintBurnMaxTokenQuantity
-    when
-        ( isJust mintingBurning' &&
-          L.any assetQuantityOutOfBounds (NE.toList $ fromJust mintingBurning')
-        ) $ liftHandler $
-            throwE ErrConstructTxMintOrBurnAssetQuantityOutOfBounds
-
-    let checkIx (ApiStakeKeyIndex (ApiT derIndex)) =
-            derIndex == DerivationIndex (getIndex @'Hardened minBound)
-    let validApiDelAction = \case
-            Joining _ stakeKeyIx -> checkIx stakeKeyIx
-            Leaving stakeKeyIx -> checkIx stakeKeyIx
-    let notall0Haccount = case body ^. #delegations of
-            Nothing -> False
-            Just delegs -> not . all validApiDelAction $ NE.toList delegs
-    when notall0Haccount $
-        liftHandler $ throwE ErrConstructTxMultiaccountNotSupported
+    apiLayer genChange knownPools getPoolStatus apiWalletId body = do
+    body & \(ApiConstructTransactionData _ _ _ _ _ _ _) ->
+    -- Above is the way to get a compiler error when number of fields changes,
+    -- in order not to forget to update the pattern below:
+        case body of
+            ApiConstructTransactionData
+                { payments = Nothing
+                , withdrawal = Nothing
+                , metadata = Nothing
+                , mintBurn = Nothing
+                , delegations = Nothing
+                } -> liftHandler $ throwE ErrConstructTxWrongPayload
+            _ -> pure ()
```

Comment:
I remember the discussion about this in the #haskell channel!

I wonder if there's a way to extract this out as a general pattern?

On the one hand, it seems desirable that when we introduce a new field, we get a compilation failure on any record pattern match that's intended to be complete. On the other hand, if we updated all similar record pattern matches across the code base, it would introduce a lot of boilerplate.

Perhaps this is something that can only be fixed in GHC.

Reply:
> I wonder if there's a way to extract this out as a general pattern?

I don't see how.

> On the other hand, if we updated all similar record pattern matches across the code base, it would introduce a lot of boilerplate.

I'd argue then that boilerplate (it is syntactically more verbose than desired) is justified by the additional guarantees that it gives.